### DISCUSSION TITLE
(뉴비주의)마우스 클릭 이벤트

### DISCUSSION THREAD

**Discussion by sudaengman** [naverz/zepeto-studio-kor#670] [5. Q&A]
대화창을 구현하고 싶어요. 최종적으로 상호작용을 하는방법을 마스터하고 싶어요.

스마트폰 환경에서 NPC에게 손으로 터치를 하면 대화창이 뜨게 하고 싶은데요

열심히 검색한 결과 그걸 구현할 함수를 찾아봤어요. ->> Input.GetMouseButton(0)

(// 뒤의 한글은 이벤트, 함수의 뜻이에요.)

    private OnTriggerEnter(other: Collider) :void  // 엔피씨에 닿으면 다음 이벤트를 실행하겠다는 이벤트예요.
    {
        
        if (other.tag == "ZepetoCharacter_" || "ZepetoPlayer")  // 제페토 캐릭터가 엔피씨에 닿는다면
        {
            if (Input.GetMouseButton(0)){      // 제페토캐릭터가 엔피씨에 닿아있는 상태에서 엔피씨에게 클릭을 한다면
                this.main.GetComponent<MainScript>().NPCChatEnter(this.text);    // 대화창을 띄우겠다.
                console.log("호우우우우우 ㅋㅋㅋㅋ2");  // 잘 작동하는지 검증하는 콘솔로그용
            }
        }
    }

근데... 테스트할때 클릭 해도 대화창도 안뜨고 콘솔로그도 안떠서 당황스러운데.. 제페토 내 대화창을 띄울 때 따로 사용되는 함수나 이벤트가 있나요...? ;ㅇ;... 독학 왕초보인데 너무 매워요........ OnTriggerEnter와  if (other.tag == "ZepetoCharacter_" || "ZepetoPlayer") 까지는 실행이 되는데  if (Input.GetMouseButton(0))는 실행이 안 되는 상황이에요..

![image](https://user-images.githubusercontent.com/102212003/165493555-c25da874-63e3-4927-9e9b-3d5e8d51edd0.png)
(전체 코드예요.. if문 안에서 작동이 안되니까 그냥 바깥에다가도 빼보고 이거저것 시도해봤는데 클릭해도 뭐도 안뜨는..ㅜㅜㅜ)

![image](https://user-images.githubusercontent.com/102212003/165495092-4501e5bd-2459-4727-a75d-b573b8d0afd5.png)
닿으면 대화창이 뜨는 NPC입니다

![image](https://user-images.githubusercontent.com/102212003/165495217-4e76ddc8-3dd6-48f1-91db-161dc6a7623c.png)
닿아서 클릭해도 대화창이 뜨지않은 문제의NPC...

**Comment by Zepeto-TechEvangelist** [comment]
안녕하세요, GetMouseClick 이벤트를 계속 감지하도록 Update()안에 구현해 보세요. 
아래는 예시입니다.

Update() {
        this.MouseControl();
}
    

MouseControl()
{
        
        if(Input.GetMouseButtonDown(0))
        {
            //NPC대화창 호출
        }
}

**Reply by sudaengman** [reply]
어??? 까닭은 모르겠지만.... 질문 몇가지 더 남겨도 될까요..?
1. GetMouseClick 이벤트와 MouseControl() 함수는 UnityEngine에 속해있는 녀석들인가요?
2. Update()에 넣는 이유는..? Update란 무엇이고 무슨역할을 하는지..?
타입스크립트 C#까지 아예 초보라서 ... 답변 남겨주시면 정말 감사드립니다.ㅜㅜㅜㅜ

**Reply by Zepeto-TechEvangelist** [reply]
1. GetMouseClick은 UnityEngine에 속해있으며 상단에 import 한 Input 으로 인해 사용할 수 있습니다.
MouseControl은 별도로 선언한 함수입니다. 
Update내에 바로 if(Input.GetMouseButtonDown(0))을 작성해도 되지만,
나중에 기능이 많아질 경우 각 기능별로 관리를 위해서는 별도의 함수를 작성하는 방식을 추천합니다.😄

2. Update는 매 프레임마다 실행되는 함수입니다. 일반적인 프로그래밍에서 무한반복(loop) 영역에 속한다고 보시면 됩니다.
세부적으로 들어가면 더 다양한 life cycle이 있지만 unity의 가장 기초적인 두 life cycle이
* Start() : play를 눌렀을때 한번 실행되는 영역 - 주로 초기화나 불러오기에 사용됩니다. 
* Update() : 계속 실행되어야 하는 코드가 들어가는 영역 - 샘플 예제에서 큐브가 계속 돌아가는 것도 Update 영역 안에 rotate코드가 들어있어서 입니다.

**Reply by sudaengman** [reply]
감사합니다 !!! 마우스클릭이벤트는 반복하는곳인 Update()안에 들어가야 되는거군요~ 만약 MouseControl()을 Update에서 빼면 어떤일이 벌어지는지는 직접한번 해볼게요! 덕분에 구현이 되었습니다~~~
